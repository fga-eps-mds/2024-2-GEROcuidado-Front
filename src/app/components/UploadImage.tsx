import React, { useState } from "react";
import { Image, View, StyleSheet, Pressable } from "react-native";
import * as ImagePicker from "expo-image-picker";
import Icon from "react-native-vector-icons/MaterialCommunityIcons";

interface Props {
  setFoto: (foto: string) => void;
  uri?: string | null;
}

export default function UploadImage({ setFoto, uri = null }: Readonly<Props>) {
  const [image, setImage] = useState<string | undefined | null>(uri);

  const pickImage = () => {
    ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [1, 1],
      base64: true,
      quality: 0,
    }).then((result) => {
      if (result.canceled) return;

      setImage(result.assets[0].uri);
      setFoto(result.assets[0].base64 as string);
    });
  };

  return (
    <View style={styles.foto}>
      <Icon style={styles.icone} name="image-outline" size={20} />
      <Pressable
        style={styles.botao}
        onPress={pickImage}
        testID="upload-image-botao"
      />
      {image && <Image source={{ uri: image }} style={styles.imagem} />}
    </View>
  );
}

const styles = StyleSheet.create({
  imagem: {
    position: "absolute",
    width: 170,
    height: 170,
    zIndex: 2,
    borderRadius: 25,
  },
  foto: {
    position: "relative",
    backgroundColor: "#EFEFF0",
    borderRadius: 25,
    alignItems: "center",
    display: "flex",
    width: 170,
    height: 170,
    alignSelf: "center",
    borderWidth: 1,
    borderColor: "#AFB1B6",
    marginBottom: 38,
  },
  botao: {
    width: 167,
    height: 174,
    backgroundColor: "transparent",
    zIndex: 3,
  },
  icone: {
    position: "absolute",
    right: "44%",
    bottom: "44%",
    opacity: 0.4,
    margin: "auto",
    alignSelf: "center",
    zIndex: 1,
  },
});
