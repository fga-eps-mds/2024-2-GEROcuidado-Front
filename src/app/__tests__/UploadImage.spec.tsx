import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import UploadImage from "../components/UploadImage";

// Mock do ImagePicker
jest.mock("expo-image-picker", () => ({
  launchImageLibraryAsync: jest.fn().mockResolvedValue({
    cancelled: false,
    assets: [{ uri: "mocked-image-uri", base64: "mocked-base64-data" }],
  }),
  MediaTypeOptions: {
    All: "All",
  },
}));

describe("UploadImage", () => {
  it("renderiza corretamente sem imagem inicial", () => {
    const { queryByTestId } = render(<UploadImage setFoto={() => {}} />);
    const imagemElement = queryByTestId("upload-image-imagem");
    expect(imagemElement).toBeNull();
  });
  it("deve renderizar corretamente", () => {
    const setFoto = jest.fn();
    const { toJSON } = render(<UploadImage setFoto={setFoto} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("escolher imagem e chamar setFoto corretamente", async () => {
    const setFotoMock = jest.fn();
    const { getByTestId } = render(<UploadImage setFoto={setFotoMock} />);

    const pickImageButton = getByTestId("upload-image-botao");

    // Simule o clique no botão para acionar a função pickImage
    fireEvent.press(pickImageButton);

    // Aguarde a resolução da promessa após a seleção da imagem
    await waitFor(() => {
      expect(setFotoMock).toHaveBeenCalled();
      expect(setFotoMock).toHaveBeenCalledWith("mocked-base64-data");
    });
  });
});
