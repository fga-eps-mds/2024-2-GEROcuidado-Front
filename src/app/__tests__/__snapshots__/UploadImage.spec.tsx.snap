// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UploadImage deve renderizar corretamente 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "alignSelf": "center",
      "backgroundColor": "#EFEFF0",
      "borderColor": "#AFB1B6",
      "borderRadius": 25,
      "borderWidth": 1,
      "display": "flex",
      "height": 170,
      "marginBottom": 38,
      "position": "relative",
      "width": 170,
    }
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": undefined,
          "fontSize": 20,
        },
        {
          "alignSelf": "center",
          "bottom": "44%",
          "margin": "auto",
          "opacity": 0.4,
          "position": "absolute",
          "right": "44%",
          "zIndex": 1,
        },
        {
          "fontFamily": "material-community",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
  >
    ó°¥¶
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "transparent",
        "height": 174,
        "width": 167,
        "zIndex": 3,
      }
    }
    testID="upload-image-botao"
  />
</View>
`;
